ipaserver/plugins/agreement.py

```diff
@@ -0,0 +1,151 @@
+from ipalib import api, errors, output
+from ipalib import Bool, Str, StrEnum
+from ipalib.plugable import Registry
+from ipaserver.plugins.baseldap import (
+    LDAPObject, LDAPSearch, LDAPCreate, LDAPDelete,
+    LDAPUpdate, LDAPRetrieve, LDAPAddMember, LDAPRemoveMember,
+    global_output_params, pkey_to_value)
+from ipaserver.plugins.hbacrule import is_all
+from ipalib import _, ngettext
+from ipapython.dn import DN
+
+
+__doc__ = _("""
+""")
+
+register = Registry()
+
+
+@register()
+class agreement(LDAPObject):
+    """
+    User Agreement object.
+    """
+    container_dn = DN(("cn", "agreements"))
+    object_name = _('Agreement')
+    object_name_plural = _('Agreements')
+    object_class = ['ipaassociation', 'fasagreement']
+    permission_filter_objectclasses = ['fasagreement']
+    default_attributes = [
+        'cn', 'description', 'member',
+    ]
+    uuid_attribute = 'ipauniqueid'
+    attribute_members = {
+        'memberuser': ['user'],
+    }
+    managed_permissions = {
+        'System: Read FAS Agreements': {
+            'replaces_global_anonymous_aci': True,
+            'ipapermbindruletype': 'all',
+            'ipapermright': {'read', 'search', 'compare'},
+            'ipapermdefaultattr': {
+                'cn', 'description', 'ipauniqueid',
+                'objectclass', 'member',
+            },
+        },
+        'System: Add FAS Agreement': {
+            'ipapermright': {'add'},
+            'default_privileges': {'CA Administrator'},
+        },
+        'System: Delete FAS Agreement': {
+            'ipapermright': {'delete'},
+            'default_privileges': {'CA Administrator'},
+        },
+        'System: Manage FAS Agreement Membership': {
+            'ipapermright': {'write'},
+            'ipapermdefaultattr': {
+                'ipacacategory', 'ipamemberca',
+                'ipacertprofilecategory', 'ipamembercertprofile',
+                'usercategory', 'memberuser',
+                'hostcategory', 'memberhost',
+                'servicecategory', 'memberservice'
+            },
+            'default_privileges': {'CA Administrator'},
+        },
+        'System: Modify FAS Agreement': {
+            'ipapermright': {'write'},
+            'ipapermdefaultattr': {
+                'cn', 'description', 'ipaenabledflag',
+            },
+            'default_privileges': {'CA Administrator'},
+        },
+    }
+
+    label = _('User Agreements')
+    label_singular = _('User Agreement')
+
+    takes_params = (
+        Str('cn',
+            cli_name='name',
+            label=_('ACL name'),
+            primary_key=True,
+        ),
+        Str('description?',
+            cli_name='desc',
+            label=_('Description'),
+        ),
+        Str('memberuser_user?',
```

Comment:
I don't think we want this parameter. Members are added/removed with ``agreement_add_user`` and ``agreement_remove_user``.